File: dkplayer-java/src/main/java/xyz/doikki/videoplayer/controller/GestureVideoController.java
package xyz.doikki.videoplayer.controller;

import android.app.Activity;
import android.content.Context;
import android.media.AudioManager;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Map;

import xyz.doikki.videoplayer.controller.component.ControlComponent;
import xyz.doikki.videoplayer.controller.component.GestureControlComponent;
import xyz.doikki.videoplayer.render.ScreenMode;
import xyz.doikki.videoplayer.util.PlayerUtils;

/**
 * 包含手势操作的VideoController
 * Created by Doikki on 2018/1/6.
 */

public abstract class GestureVideoController extends MediaController implements
        GestureDetector.OnGestureListener,
        GestureDetector.OnDoubleTapListener,
        View.OnTouchListener {

    private GestureDetector mGestureDetector;
    private AudioManager mAudioManager;
    private boolean mIsGestureEnabled = true;
    private int mStreamVolume;
    private float mBrightness;
    private int mSeekPosition = -1;
    private boolean mFirstTouch;
    private boolean mChangePosition;
    private boolean mChangeBrightness;
    private boolean mChangeVolume;

    private boolean mCanChangePosition = true;

    private boolean mEnableInNormal;

    private boolean mCanSlide;



    private boolean mIsDoubleTapTogglePlayEnabled = true;


    public GestureVideoController(@NonNull Context context) {
        super(context);
    }

    public GestureVideoController(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public GestureVideoController(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void initView() {
        super.initView();
        mAudioManager = (AudioManager) getContext().getSystemService(Context.AUDIO_SERVICE);
        mGestureDetector = new GestureDetector(getContext(), this);
        setOnTouchListener(this);
    }

    /**
     * 设置是否可以滑动调节进度，默认可以
     */
    public void setCanChangePosition(boolean canChangePosition) {
        mCanChangePosition = canChangePosition;
    }

    /**
     * 是否在竖屏模式下开始手势控制，默认关闭
     */
    public void setEnableInNormal(boolean enableInNormal) {
        mEnableInNormal = enableInNormal;
    }

    /**
     * 是否开启手势控制，默认开启，关闭之后，手势调节进度，音量，亮度功能将关闭
     */
    public void setGestureEnabled(boolean gestureEnabled) {
        mIsGestureEnabled = gestureEnabled;
    }

    /**
     * 是否开启双击播放/暂停，默认开启
     */
    public void setDoubleTapTogglePlayEnabled(boolean enabled) {
        mIsDoubleTapTogglePlayEnabled = enabled;
    }

    @Override
    public void setScreenMode(int playerState) {
        super.setScreenMode(playerState);
        if (playerState == ScreenMode.NORMAL) {
            mCanSlide = mEnableInNormal;
        } else if (playerState == ScreenMode.FULL) {
            mCanSlide = true;
        }
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        return mGestureDetector.onTouchEvent(event);
    }

    /**
     * 手指按下的瞬间
     */
    @Override
    public boolean onDown(MotionEvent e) {
        if (!isInPlaybackState() //不处于播放状态
                || !mIsGestureEnabled //关闭了手势
                || PlayerUtils.isEdge(getContext(), e)) //处于屏幕边沿
            return true;
        mStreamVolume = mAudioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
        Activity activity = PlayerUtils.scanForActivity(getContext());
        if (activity == null) {
            mBrightness = 0;
        } else {
            mBrightness = activity.getWindow().getAttributes().screenBrightness;
        }
        mFirstTouch = true;
        mChangePosition = false;
        mChangeBrightness = false;
        mChangeVolume = false;
        return true;
    }

    /**
     * 单击
     */
    @Override
    public boolean onSingleTapConfirmed(MotionEvent e) {
        if (isInPlaybackState()) {
            mControlWrapper.toggleShowState();
        }
        return true;
    }

    /**
     * 双击
     */
    @Override
    public boolean onDoubleTap(MotionEvent e) {
        if (mIsDoubleTapTogglePlayEnabled && !isLocked() && isInPlaybackState()) togglePlay();
        return true;
    }

    /**
     * 在屏幕上滑动
     */
    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        if (!isInPlaybackState() //不处于播放状态
                || !mIsGestureEnabled //关闭了手势
                || !mCanSlide //关闭了滑动手势
                || isLocked() //锁住了屏幕
                || PlayerUtils.isEdge(getContext(), e1)) //处于屏幕边沿
            return true;
        float deltaX = e1.getX() - e2.getX();
        float deltaY = e1.getY() - e2.getY();
        if (mFirstTouch) {
            mChangePosition = Math.abs(distanceX) >= Math.abs(distanceY);
            if (!mChangePosition) {
                //半屏宽度
                int halfScreen = PlayerUtils.getScreenWidth(getContext(), true) / 2;
                if (e2.getX() > halfScreen) {
                    mChangeVolume = true;
                } else {
                    mChangeBrightness = true;
                }
            }

            if (mChangePosition) {
                //根据用户设置是否可以滑动调节进度来决定最终是否可以滑动调节进度
                mChangePosition = mCanChangePosition;
            }

            if (mChangePosition || mChangeBrightness || mChangeVolume) {
                for (Map.Entry<ControlComponent, Boolean> next : mControlComponents.entrySet()) {
                    ControlComponent component = next.getKey();
                    if (component instanceof GestureControlComponent) {
                        ((GestureControlComponent) component).onStartSlide();
                    }
                }
            }
            mFirstTouch = false;
        }
        if (mChangePosition) {
            slideToChangePosition(deltaX);
        } else if (mChangeBrightness) {
            slideToChangeBrightness(deltaY);
        } else if (mChangeVolume) {
            slideToChangeVolume(deltaY);
        }
        return true;
    }

    protected void slideToChangePosition(float deltaX) {
        deltaX = -deltaX;
        int width = getMeasuredWidth();
        int duration = (int) mControlWrapper.getDuration();
        int currentPosition = (int) mControlWrapper.getCurrentPosition();
        int position = (int) (deltaX / width * 120000 + currentPosition);
        if (position > duration) position = duration;
        if (position < 0) position = 0;
        for (Map.Entry<ControlComponent, Boolean> next : mControlComponents.entrySet()) {
            ControlComponent component = next.getKey();
            if (component instanceof GestureControlComponent) {
                ((GestureControlComponent) component).onPositionChange(position, currentPosition, duration);
            }
        }
        mSeekPosition = position;
    }

    protected void slideToChangeBrightness(float deltaY) {
        Activity activity = PlayerUtils.scanForActivity(getContext());
        if (activity == null) return;
        Window window = activity.getWindow();
        WindowManager.LayoutParams attributes = window.getAttributes();
        int height = getMeasuredHeight();
        if (mBrightness == -1.0f) mBrightness = 0.5f;
        float brightness = deltaY * 2 / height + mBrightness;
        if (brightness < 0) {
            brightness = 0f;
        }
        if (brightness > 1.0f) brightness = 1.0f;
        int percent = (int) (brightness * 100);
        attributes.screenBrightness = brightness;
        window.setAttributes(attributes);
        for (Map.Entry<ControlComponent, Boolean> next : mControlComponents.entrySet()) {
            ControlComponent component = next.getKey();
            if (component instanceof GestureControlComponent) {
                ((GestureControlComponent) component).onBrightnessChange(percent);
            }
        }
    }

    protected void slideToChangeVolume(float deltaY) {
        int streamMaxVolume = mAudioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        int height = getMeasuredHeight();
        float deltaV = deltaY * 2 / height * streamMaxVolume;
        float index = mStreamVolume + deltaV;
        if (index > streamMaxVolume) index = streamMaxVolume;
        if (index < 0) index = 0;
        int percent = (int) (index / streamMaxVolume * 100);
        mAudioManager.setStreamVolume(AudioManager.STREAM_MUSIC, (int) index, 0);
        for (Map.Entry<ControlComponent, Boolean> next : mControlComponents.entrySet()) {
            ControlComponent component = next.getKey();
            if (component instanceof GestureControlComponent) {
                ((GestureControlComponent) component).onVolumeChange(percent);
            }
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        //滑动结束时事件处理
        if (!mGestureDetector.onTouchEvent(event)) {
            int action = event.getAction();
            switch (action) {
                case MotionEvent.ACTION_UP:
                    stopSlide();
                    if (mSeekPosition >= 0) {
                        mControlWrapper.seekTo(mSeekPosition);
                        mSeekPosition = -1;
                    }
                    break;
                case MotionEvent.ACTION_CANCEL:
                    stopSlide();
                    mSeekPosition = -1;
                    break;
            }
        }
        return super.onTouchEvent(event);
    }

    private void stopSlide() {
        for (Map.Entry<ControlComponent, Boolean> next : mControlComponents.entrySet()) {
            ControlComponent component = next.getKey();
            if (component instanceof GestureControlComponent) {
                ((GestureControlComponent) component).onStopSlide();
            }
        }
    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        return false;
    }

    @Override
    public void onLongPress(MotionEvent e) {

    }

    @Override
    public void onShowPress(MotionEvent e) {

    }

    @Override
    public boolean onDoubleTapEvent(MotionEvent e) {
        return false;
    }


    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        return false;
    }
}

File: dkplayer-java/src/main/java/xyz/doikki/videoplayer/controller/GestureVideoController.kt
package xyz.doikki.videoplayer.controller

import android.content.Context
import android.media.AudioManager
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import androidx.annotation.AttrRes
import xyz.doikki.videoplayer.VideoView
import xyz.doikki.videoplayer.util.INVALIDATE_SEEK_POSITION
import xyz.doikki.videoplayer.controller.component.GestureControlComponent
import xyz.doikki.videoplayer.util.getActivityContext
import xyz.doikki.videoplayer.util.PlayerUtils
import kotlin.math.abs

/**
 * 包含手势操作的VideoController
 * Created by Doikki on 2018/1/6.
 */
abstract class GestureVideoController @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, @AttrRes defStyleAttr: Int = 0
) : VideoController(context, attrs, defStyleAttr),
    GestureDetector.OnGestureListener,
    GestureDetector.OnDoubleTapListener,
    OnTouchListener {

    private val gestureDetector by lazy {
        GestureDetector(context, this)
    }
    private val audioManager by lazy {
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

    /**
     * 是否在竖屏模式下开始手势控制，默认关闭
     */
    var gestureInPortraitEnabled = false

    /**
     * 是否开启手势控制，默认开启，关闭之后，手势调节进度，音量，亮度功能将关闭
     */
    var gestureEnabled = true

    /**
     * 是否开启双击播放/暂停，默认开启
     */
    var doubleTapTogglePlayEnabled = true

    /**
     * 设置是否可以滑动调节进度，默认可以
     */
    var seekEnabled = true

    //是否可以滑动：滑动调节音量或者亮度
    private var canSlide = false

    //待处理的seek position：通常由于手势滑动或者按键引起的位置变动
    protected var pendingSeekPosition: Int = INVALIDATE_SEEK_POSITION

    private var streamVolume = 0
    private var brightness = 0f

    private var firstTouch = false
    private var changePosition = false
    private var changeBrightness = false
    private var changeVolume = false

    init {
        setOnTouchListener(this)
    }

    override fun setScreenMode(@VideoView.ScreenMode screenMode: Int) {
        super.setScreenMode(screenMode)
        if (screenMode == VideoView.SCREEN_MODE_NORMAL) {
            canSlide = gestureInPortraitEnabled
        } else if (screenMode == VideoView.SCREEN_MODE_FULL) {
            canSlide = true
        }
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        return gestureDetector.onTouchEvent(event)
    }

    /**
     * 手指按下的瞬间
     */
    override fun onDown(e: MotionEvent): Boolean {
        if (!isInPlaybackState //不处于播放状态
            || !gestureEnabled //关闭了手势
            || PlayerUtils.isEdge(context, e) //处于屏幕边沿
        )
            return true
        streamVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)
        val activity = context.getActivityContext()
        brightness = activity?.window?.attributes?.screenBrightness ?: 0f
        firstTouch = true
        changePosition = false
        changeBrightness = false
        changeVolume = false
        return true
    }

    override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
        if (isInPlaybackState) {
            toggleShowState()
        }
        return true
    }

    /**
     * 双击
     */
    override fun onDoubleTap(e: MotionEvent): Boolean {
        if (doubleTapTogglePlayEnabled && !isLocked && isInPlaybackState) togglePlay()
        return true
    }

    /**
     * 在屏幕上滑动
     */
    override fun onScroll(
        e1: MotionEvent,
        e2: MotionEvent,
        distanceX: Float,
        distanceY: Float
    ): Boolean {
        if (!isInPlaybackState //不处于播放状态
            || !gestureEnabled //关闭了手势
            || !canSlide //关闭了滑动手势
            || isLocked //锁住了屏幕
            || PlayerUtils.isEdge(context, e1) //处于屏幕边沿
        )
            return true
        val deltaX = e1.x - e2.x
        val deltaY = e1.y - e2.y
        if (firstTouch) {
            changePosition = abs(distanceX) >= abs(distanceY)
            if (!changePosition) {
                //半屏宽度
                val halfScreen = PlayerUtils.getScreenWidth(context, true) / 2
                if (e2.x > halfScreen) {
                    changeVolume = true
                } else {
                    changeBrightness = true
                }
            }
            if (changePosition) {
                //根据用户设置是否可以滑动调节进度来决定最终是否可以滑动调节进度
                changePosition = seekEnabled
            }
            if (changePosition || changeBrightness || changeVolume) {
                for ((component) in controlComponents) {
                    if (component is GestureControlComponent) {
                        component.onStartSlide()
                    }
                }
            }
            firstTouch = false
        }
        if (changePosition) {
            slideToChangePosition(deltaX)
        } else if (changeBrightness) {
            slideToChangeBrightness(deltaY)
        } else if (changeVolume) {
            slideToChangeVolume(deltaY)
        }
        return true
    }

    /**
     * 滑动切换播放位置
     */
    protected fun slideToChangePosition(deltaX: Float) {
        invokeOnPlayerAttached {
            val width = measuredWidth
            val duration = it.duration.toInt()
            val currentPosition = it.currentPosition.toInt()
            var position = (-deltaX / width * 120000 + currentPosition).toInt()
            if (position > duration) position = duration
            if (position < 0) position = 0
            setPendingSeekPositionAndNotify(position, currentPosition, duration)
        }
    }


    protected fun slideToChangeBrightness(deltaY: Float) {
        val activity = PlayerUtils.scanForActivity(context) ?: return
        val window = activity.window
        val attributes = window.attributes
        val height = measuredHeight
        if (brightness == -1.0f) brightness = 0.5f
        var brightness = deltaY * 2 / height + brightness
        if (brightness < 0) {
            brightness = 0f
        }
        if (brightness > 1.0f) brightness = 1.0f
        val percent = (brightness * 100).toInt()
        attributes.screenBrightness = brightness
        window.attributes = attributes
        for ((component) in controlComponents) {
            if (component is GestureControlComponent) {
                component.onBrightnessChange(percent)
            }
        }
    }

    protected fun slideToChangeVolume(deltaY: Float) {
        val streamMaxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        val height = measuredHeight
        val deltaV = deltaY * 2 / height * streamMaxVolume
        var index = streamVolume + deltaV
        if (index > streamMaxVolume) index = streamMaxVolume.toFloat()
        if (index < 0) index = 0f
        val percent = (index / streamMaxVolume * 100).toInt()
        audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, index.toInt(), 0)
        for ((component) in controlComponents) {
            if (component is GestureControlComponent) {
                component.onVolumeChange(percent)
            }
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        //滑动结束时事件处理
        if (!gestureDetector.onTouchEvent(event)) {
            when (event.action) {
                MotionEvent.ACTION_UP -> {
                    stopSlide()
                    handlePendingSeekPosition()
                }
                MotionEvent.ACTION_CANCEL -> {
                    stopSlide()
                    cancelPendingSeekPosition()
                }
            }
        }
        return super.onTouchEvent(event)
    }

    protected fun setPendingSeekPositionAndNotify(
        position: Int,
        currentPosition: Int,
        duration: Int
    ) {
        for ((component) in controlComponents) {
            if (component is GestureControlComponent) {
                component.onPositionChange(position, currentPosition, duration)
            }
        }
        pendingSeekPosition = position
    }

    protected open fun handlePendingSeekPosition() {
        invokeOnPlayerAttached { player ->
            if (pendingSeekPosition >= 0) {
                player.seekTo(pendingSeekPosition.toLong())
            }
        }
        pendingSeekPosition = INVALIDATE_SEEK_POSITION
    }

    protected open fun cancelPendingSeekPosition() {
        pendingSeekPosition = INVALIDATE_SEEK_POSITION
    }

    private fun stopSlide() {
        for ((component) in controlComponents) {
            if (component is GestureControlComponent) {
                component.onStopSlide()
            }
        }
    }

    override fun onFling(
        e1: MotionEvent,
        e2: MotionEvent,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        return false
    }

    override fun onLongPress(e: MotionEvent) {}
    override fun onShowPress(e: MotionEvent) {}
    override fun onDoubleTapEvent(e: MotionEvent): Boolean {
        return false
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean {
        return false
    }
}